presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java

```diff
@@ -370,6 +382,95 @@ public void testHashBuilderResize(boolean hashEnabled, long memoryLimitForMerge,
+        // each hash table entry is of 13 bytes and the fill ratio is 0.75;
+        // so we need 1_000_000 / 13 * 0.75 = 58K entries for the first page in order to exceed 1MB limit
+        // set the initial length to be 1 << 16 to exceed 58K entries
+        int length = 1 << 16;
+        int expected = 0;
+        for (int i = 0; i < 5; i++) {
+            expected += length;
+            rowPagesBuilder.addSequencePage(length, length);
+            length = length * 2;
+        }
+        List<Page> input = rowPagesBuilder.build();
+
+        // mock an adjustable memory pool
+        QueryId queryId = new QueryId("test_query");
+        MemoryPool memoryPool = new MemoryPool(new MemoryPoolId("test"), new DataSize(1, GIGABYTE));
+        QueryContext queryContext = new QueryContext(
+                queryId,
+                new DataSize(512, MEGABYTE),
+                memoryPool,
+                new MemoryPool(new MemoryPoolId("test-system"), new DataSize(512, MEGABYTE)),
+                executor,
+                scheduledExecutor,
+                new DataSize(512, MEGABYTE),
+                new SpillSpaceTracker(new DataSize(512, MEGABYTE)));
+
+        DriverContext driverContext = createTaskContext(queryContext, executor, TEST_SESSION)
+                .addPipelineContext(0, true, true)
+                .addDriverContext();
+
+        HashAggregationOperatorFactory operatorFactory = new HashAggregationOperatorFactory(
+                0,
+                new PlanNodeId("test"),
+                ImmutableList.of(VARCHAR),
+                hashChannels,
+                ImmutableList.of(),
+                Step.SINGLE,
+                ImmutableList.of(COUNT.bind(ImmutableList.of(0), Optional.empty())),
+                rowPagesBuilder.getHashChannel(),
+                Optional.empty(),
+                1_000,
+                new DataSize(16, MEGABYTE),
+                joinCompiler);
+
+        // test yield by setting memory pool free bytes
+        Operator operator = operatorFactory.createOperator(driverContext);
+        for (Page page : input) {
+            // unblocked
+            assertTrue(operator.needsInput());
+
+            // saturate the pool
+            memoryPool.reserve(queryId, 2_000_000_000);
+
+            long memoryUsage = operator.getOperatorContext().getDriverContext().getMemoryUsage();
+            operator.addInput(page);
+            long increasedMemoryUsage = operator.getOperatorContext().getDriverContext().getMemoryUsage() - memoryUsage;
+            // assert we didn't allocate a large hash table
+            // each entry in the hash table is with a long, integer, and byte type, which is 13 bytes totally
+            // the hash table also has a fill ratio of 0.75
+            assertLessThan(increasedMemoryUsage, (Long.BYTES + Integer.BYTES + Byte.BYTES) * nextPowerOfTwo((long) (page.getPositionCount() / 0.75)));
```

Comment:
This is not sufficient.

In-person discussion is necessary:
* How is memory for Work object accounted (specifically, the Page in it)?
* Give plausible explanation for memory usage after each invocation.
* Make the test simpler and more effective. (reduce math would be one step)